회원 계정 관련 기본 기능 구현

zhdiddl opened the pull request:
**변경 사항**
- MySQL 서버 설정 및 인증 관련 기능 의존성을 추가했습니다.
- 회원 계정 CRUD 기능을 추가했습니다.
- Spring Security 사용해서 인증 정보를 로드할 수 있도록 설정했습니다.
- 서버 부담을 감소하고 확장성을 높이기 위해 토큰 기반 인증 방식으로 구현했습니다.
- 사용자 환경에 맞는 인증 방식을 제공할 수 있도록 헤더와 쿠키 저장 옵션을 제공했습니다.

**참고 사항**
- 유저 계정에 역할 분류 기능이 추가되지 않은 상태입니다.
- 유저 계정 삭제에 소프트 삭제를 적용하면서 isDeleted 필드가 추가되었습니다.
- 가입시 이메일 인증 기능이 추가되지 않은 상태입니다.
- Swagger를 통해 API 테스트가 가능합니다.

zhdiddl committed:
feat: 회원 계정 관련 기본 기능 구현

- MySQL 서버 설정 및 인증 관련 기능 의존성을 추가했습니다.
- 회원 계정 CRUD 기능을 추가했습니다.
- Spring Security 사용해서 인증 정보를 로드할 수 있도록 설정했습니다.
- 서버 부담을 감소하고 확장성을 높이기 위해 토큰 기반 인증 방식으로 구현했습니다.
- 사용자 환경에 맞는 인증 방식을 제공할 수 있도록 헤더와 쿠키 저장 옵션을 제공했습니다.

spring-kotlin-lover commented on src/main/java/com/learnhive/lessonservice/controller/UserAccountController.java:
```diff
@@ -0,0 +1,120 @@
+    private final JwtBlacklistService jwtBlacklistService;
+
+    @PostMapping("/signUp")
+    public ResponseEntity<String> signUp(@Valid @RequestBody UserAccountDto userForm, BindingResult result) {
+        // 유효성 검증 실패 시
+        if (result.hasErrors()) {
+            return ResponseEntity.badRequest().body(result.getAllErrors().stream()
+                    .map(ObjectError::getDefaultMessage).collect(Collectors.joining(", ")));
+        }
```
유효성 검증시 실패하는 경우에 대한 로직을 `ExceptionController` 쪽으로 옮겨보는 것은 어떨까요? 

전반적으로 유효성 검증 로직은 controller 로직을 길게 가져가는 것에 단점이 있고, 이러한 코드들이 반복된다는 것이 문제입니다.
`MethodArgumentNotValidException` 을 잘 핸들링하면 로직을 더 심플하게 가져갈 수 있으니 개선해보면 좋겟네요.

https://velog.io/@dangddoong/Spring-Valid-Exception-Handling-MethodArgumentNotValidException

spring-kotlin-lover commented on src/main/java/com/learnhive/lessonservice/controller/UserAccountController.java:
```diff
@@ -0,0 +1,120 @@
+    @PutMapping("/{userId}")
+    public ResponseEntity<String> updateUser(@Parameter(description = "정보를 변경할 유저 아이디", required = true)
+                                             @PathVariable Long userId,
+                                             @RequestBody UserAccountDto updateForm) {
```
UserAccountDto 에도 `@Valid` 어노테이션을 추가하는 것은 어떨까요?

spring-kotlin-lover commented on src/main/java/com/learnhive/lessonservice/controller/UserAccountController.java:
```diff
@@ -0,0 +1,120 @@
+    public ResponseEntity<String> deleteUser(@Parameter(description = "삭제할 유저 아이디", required = true)
+                                             @PathVariable Long userId,
+                                             HttpServletResponse response) {
+        userService.deleteUser(userId, response);
```
service layer 로 HttpServeletResponse 를 argument 로 넘기지 말아야될 것 같습니다. service layer 는 통신 방법과 상관없는 영역이라서, Http 인지 grpc 인지 통신 방식에 대해서는 관심이 없어야 합니다.

설계적으로 조금 더 개선해보는 것은 어떨까요?

spring-kotlin-lover commented on src/main/java/com/learnhive/lessonservice/domain/UserAccount.java:
```diff
@@ -0,0 +1,39 @@
+@AllArgsConstructor
+@NoArgsConstructor
+@Entity
+public class UserAccount extends AuditingFields{
```
```suggestion
public class UserAccount extends AuditingFields {
```
java coding convention 에 대해서는 꼭 잘지킬 수 있도록 노력해주세요 ^^;

spring-kotlin-lover commented on src/main/java/com/learnhive/lessonservice/service/UserAccountService.java:
```diff
@@ -0,0 +1,169 @@
+
+        // 소프트 삭제 처리
+        authenticatedUser.setDeleted(true);
+        userAccountRepository.save(authenticatedUser);
```
JPA 에서 제공하는 dirty check 매커니즘을 이해하고 적용해보시면 어떤가 싶네요

JPA 에서 dirty check 를 이용하면 굳이 repository 의 save 를 호출하지 않아도 됩니다. spring 에서 제공하는 transactional 어노테이션 기반으로 트랜잭션 범위를 service 로 정의하고, 데이터 속성을 변경만 하게 되면 영속성 컨텍스트에 있는 entity 정보를 기반으로 diff 를 찾아내 update 쿼리를 자동으로 jpa 에서 만들게 된답니다.

아래 링크 참고드립니다
https://jojoldu.tistory.com/415

zhdiddl committed:
refactor: 유효성 검사 로직 분리

- DTO에서 수행한 유효성 검사 수행 후 예외 발생시 ExceptionController에서 전역적으로 처리할 수 있도록 수정
- ExceptionResponse 내부 클래스가 예외 발생시 응답을 본문에 반환할 수 있도록 @getter 추가

zhdiddl committed:
chore: 자바 코딩 컨벤션에 맞게 코드 수정

zhdiddl committed:
refactor: 서비스 로직 변경에 맞게 컨트롤러 로직 수정

- 유효성 검사 관련 로직을 제거하고 전역 예외 처리 컨트롤러로 책임을 위임
- 서비스 레이어로 HttpServletResponse가 인자로 넘어가지 않도록 코드를 개선하면서 그에 맞게 컨트롤러 로직 수정
- 예외 발생 시 상세 정보를 로그에 남기도록 추가

zhdiddl commented:
@spring-kotlin-lover 멘토님 리뷰 감사합니다! 코멘트 주신 내용에 따라 코드 수정했는데 이대로 머지해도 괜찮을까요?
